helix-term/src/commands.rs

```diff
@@ -3284,33 +3286,88 @@ fn select_register(cx: &mut Context) {
     })
 }
 
-fn space_mode(cx: &mut Context) {
-    cx.on_next_key(move |cx, event| {
-        if let KeyEvent {
-            code: KeyCode::Char(ch),
-            ..
-        } = event
-        {
-            // TODO: temporarily show SPC in the mode list
-            match ch {
-                'f' => file_picker(cx),
-                'b' => buffer_picker(cx),
-                's' => symbol_picker(cx),
-                'w' => window_mode(cx),
-                'y' => yank_joined_to_clipboard(cx),
-                'Y' => yank_main_selection_to_clipboard(cx),
-                'p' => paste_clipboard_after(cx),
-                'P' => paste_clipboard_before(cx),
-                'R' => replace_selections_with_clipboard(cx),
-                // ' ' => toggle_alternate_buffer(cx),
-                // TODO: temporary since space mode took its old key
-                ' ' => keep_primary_selection(cx),
-                _ => (),
-            }
+macro_rules! mode_info {
+    // TODO: how to use one expr for both pat and expr?
+    // TODO: how to use replaced function name as str at compile time?
+    // TODO: extend to support multiple keys, but first solve the other two
+    {$name:literal, $cx:expr, $($key:expr => $func:expr; $desc:literal),+,} => {
+        mode_info! {
+            $name, $cx,
+            $($key; $key => $func; $desc,)+
         }
-    })
+    };
+    {$name:literal, $cx:expr, $($key:expr; $keyp:pat => $func:expr; $desc:literal),+,} => {
+        $cx.editor.autoinfo = Some(Info::key(
+            $name,
+            vec![
```

Comment:
Couldn't we make these definitions `const` and then define them outside of the function? Similar to how commands are handled.

Reply:
I don't think `const` works for `String`. We could, I think the next step is let me play around with docs generation a bit.

Reply:
If it's compile time it can be `&'static str`